### PR TITLE
DEVO-110:Update DHF Develop pipeline to include palamida scan

### PR DESCRIPTION
**Pull request by yafeldman**
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -636,7 +636,70 @@ pipeline{
...
+                        return (env.BRANCH_NAME==props['ExecutionBranch'])
+                    }
+                }
+            }
+            agent {label 'dhfLinuxAgent'}
+            steps{
+                script {
+
+                    def palamida_url="http://palamida-2.marklogic.com:8888"
+                    def jwt="eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJwYWxhbWlkYSIsInVzZXJJZCI6MywiaWF0IjoxNjA1NTYxMTc5fQ.5NaV_E4FHSe6vJMTX1siZa6f0U-uKm3ssZIfmsPEPzmdkDvtnNxsXRxI1Mj4mHq0i9cBNEHW_ePYKdFh97Sd9Q"
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:654]: Add these to Jenkins credentials

**Review by SameeraPriyathamTadikonda — CHANGES_REQUESTED** [review]
There's an error when running the pipeline. So extract a method out of it to avoid method too large error.

File: Jenkinsfile (hunk 1 above)
Comment (yafeldman) [reply Jenkinsfile:654]: How to add credentials in Jenkins? I tried but seem like don't have permissions.

File: Jenkinsfile [hunk 2]
```diff
@@ -229,6 +229,161 @@ def runCypressE2e(){
...
+            }
+        }}
+
+    archiveArtifacts artifacts: 'inventoriesRprt.report.json'
+
+    //Send email
+    if(failures.size()!=0){
+        def body=''
+        failures.each { msg -> body=body+msg + '\n' }
+        sendMail 'Kavitha.Sivagnanam@marklogic.com,yakov.feldman@marklogic.com', body,false,'Flexcode FAILED. HIGH vulnerabilities','inventoriesRprt.report.json'
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:276]: you don't need to send the email twice. An email will be sent in the post stage. If required include Kavitha to the recipient list.

File: Jenkinsfile [hunk 3]
```diff
@@ -229,6 +229,161 @@ def runCypressE2e(){
...
+}
+
+void PreBuildCheck() {
+    script{
+        if(env.CHANGE_ID){
+            if(PRDraftCheck()){
+                sh 'exit 1'
+            }
+            def reviewState=getReviewState()
+            if((!env.CHANGE_TITLE.startsWith("DHFPROD-"))){
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:315]: Add an OR check for DEVO project too.

File: Jenkinsfile [hunk 4]
```diff
@@ -636,7 +719,22 @@ pipeline{
                   }
                   }
 		}
-            stage('publishAnddhs'){
+        stage('FlexCodeScan'){
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:722]: These stages FlexCodeScan and PUblishAnddhs should run in parallel.

**Review by SameeraPriyathamTadikonda — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 2 above)
Comment (yafeldman) [reply Jenkinsfile:276]: Then email in `post` would need to include attached report and P1 vulnerabilities info in the body of the email. Is it OK?

Comment (SameeraPriyathamTadikonda) [reply Jenkinsfile:276]: yes

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

File: Jenkinsfile [hunk 5]
```diff
@@ -229,6 +230,152 @@ def runCypressE2e(){
...
+            env.CHANGE_URL=" "
+        }
+    }
+}
+
+void PreBuildCheck() {
+ if(env.CHANGE_ID){
+   if(PRDraftCheck()){ sh 'exit 1' }
+   def reviewState=getReviewState()
+   if((!env.CHANGE_TITLE.startsWith("DHFPROD-")) && (!env.CHANGE_TITLE.startsWith("DEVO-"))){
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:311]: Should be || instead of &&

**Review by SameeraPriyathamTadikonda — CHANGES_REQUESTED** [review]

**Review by SameeraPriyathamTadikonda — DISMISSED** [review]

File: Jenkinsfile (hunk 5 above)
Comment (yafeldman) [reply Jenkinsfile:311]: I believe it has to be && since it negative condition: 
**!**env.CHANGE_TITLE.startsWith("DHFPROD-")

File: Jenkinsfile [hunk 6]
```diff
@@ -21,17 +22,18 @@ def dhflinuxTests(String mlVersion,String type){
...
+    		sh 'docker exec -u builder -i '+dockerhost+' /bin/sh -c "su -builder;export JAVA_HOME=`eval echo "$JAVA_HOME_DIR"`;export GRADLE_USER_HOME=$WORKSPACE$GRADLE_DIR;export M2_HOME=$MAVEN_HOME/bin;export PATH=$JAVA_HOME/bin:$GRADLE_USER_HOME:$PATH:$MAVEN_HOME/bin;cd $WORKSPACE/data-hub;rm -rf $GRADLE_USER_HOME/caches;./gradlew clean;set +e;./gradlew marklogic-data-hub:bootstrapAndTest -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/;sleep 10s;./gradlew marklogic-data-hub-central:test -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/ |& tee console.log;sleep 10s;./gradlew ml-data-hub:test -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/;sleep 10s;./gradlew web:test -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/;sleep 10s;./gradlew marklogic-data-hub-spark-connector:test -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/;sleep 10s;./gradlew ml-data-hub:testFullCycle -i --stacktrace -PnodeDistributionBaseUrl=http://node-mirror.eng.marklogic.com:8080/;sleep 10s;"'
```
Comment (SameeraPriyathamTadikonda) [inline Jenkinsfile:25]: Why are these changed?

**Review by SameeraPriyathamTadikonda — DISMISSED** [review]

File: Jenkinsfile [hunk 7]
```diff
@@ -229,6 +231,154 @@ def runCypressE2e(){
         junit '**/e2e/**/*.xml'
     }
 }
+
+void postFailureFiexcodeScanAndReport(def vulnerabilities){
```
Comment (akshaysonvane) [inline Jenkinsfile:235]: Should it be Flex instead of Fiex?

File: Jenkinsfile (hunk 6 above)
Comment (yafeldman) [reply Jenkinsfile:25]: Tried to fix conflict which I am not sure why this conflict happened. Most likely came form IDE editing.

File: Jenkinsfile (hunk 7 above)
Comment (yafeldman) [reply Jenkinsfile]: fixed

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

File: Jenkinsfile [hunk 8]
```diff
@@ -229,6 +231,154 @@ def runCypressE2e(){
...
+    email = email + ',Kavitha.Sivagnanam@marklogic.com'
+
+    def body=''
+    vulnerabilities.each { msg -> body=body+msg + '<p>' }
+    body = body + '<p>scan report is attached.'
+
+    sendMail email,"<h3>Pipeline FAILED in stage ${STAGE_NAME} <a href=${BUILD_URL}/console> Check Console Output Here</a></h4>  High vulneribilities: <p> $body",false," pipeline FAILED in ${STAGE_NAME} ", 'inventoriesRprt.report.json'
+}
+
+def fiexcodeScanAndReport(){
```
Comment (akshaysonvane) [inline Jenkinsfile:252]: Need the typo fix here as well.

**Review by akshaysonvane — CHANGES_REQUESTED** [review]

**Review by akshaysonvane — APPROVED** [review]

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

File: Jenkinsfile (hunk 8 above)
Comment (yafeldman) [reply Jenkinsfile]: fixed
